### ISSUE TITLE
Open Faker for custom/community modules

### ISSUE THREAD

**Issue by xDivisionByZerox** [faker-js/faker#704]
### Clear and concise description of the problem

Faker has my different modules but even more requested features. Many of these features are so sadly too complex to be maintained by the Faker-JS team. 

### Suggested solution

The Faker class should accept a list of modules that will be used under the created Faker instance. These modules will be plugins maintained by the community.
Like:

```
const myModule = {
    test() {
        return 'test';
    }
}

const fakerWithCustomModule = new Faker({
    customModules: [
        { name: 'testModule', module: myModule },
    ],
});
fakerWithCustomModule.testModule.test(); // 'test'
```

### Additional context

Somewhat related to #444

**Comment by ST-DDT** [comment]
Here is a workaround to do this right now in a typesafe way:

```ts
import type { Faker } from "@faker-js/faker";
import { faker } from "@faker-js/faker";

function withCustomModule<
  S extends Faker,
  Name extends string,
  T,
  U extends S & Record<Name, T>
>(base: S, name: Name, value: T): U {
  return { ...base, [name]: value } as U;
}

const customModule = {
  abc: () => "sjkdfhajkshdf",
} as const;

const fakerPlus = withCustomModule(faker, "test", customModule);

console.log(fakerPlus.test.abc());

```

**Comment by ST-DDT** [comment]
Isn't that almost like my example above?
(Just with new/a constructor and Object.assign instead of a Record)

**Comment by xDivisionByZerox** [comment]
Yes, I just wanted to provide an example of how other communities handle the extendability approach.

**Comment by ST-DDT** [comment]
If you are interested in this feature, please upvote it.

**Comment by ST-DDT** [comment]
We might need something like this for fake with the standalone module functions.

I think it might end up in something like a registry:

````ts
registry.module.method
````

**Comment by matthewmayer** [comment]
I appreciate this issue has a lot of upvotes, but I feel I'm maybe missing something about the potential use-cases.

For example, suppose you really want faker to support random tree names but you notice https://github.com/faker-js/faker/issues/2022 doesn't have enough up-votes yet.

You can already easily create your own function using your own data with faker helper methods e.g. `return faker.helpers.arrayElement(["ash","oak","pine","palm" ... ])`

What is gained by making this one-line function into a "community module"? By using faker.helpers functions you already can take advantage of reproducible seeds etc. And core modules wouldn't be able to refer to community modules in their definitions. If you just want to share the data, you could already publish an NPM package which is just `module.exports = ["ash","oak","pine","palm" ... ]` and feed that into `faker.helpers.arrayElement`.

**Comment by ST-DDT** [comment]
AFAICT: The are currently 3 things this is intended for:

- make it callable via `faker.moduleA.functionB`
- make it able to use `faker.definitions.categoryC.entryD` and throw a useful error if it is missing
- make it usable via `faker.helpers.fake`

But also as a reminder, that whatever we are going to do for #2667 + #2664 it should be possible for our community to do the same for their functions/modules.

- #2667
- #2664

What I hope to do with that:

- Overwrite `faker.string.fromRegExp` with a version with dependencies supporting "full" regex (as plugin/extension).

**Comment by matthewmayer** [comment]
But what is the advantage of having it callable via `faker.moduleA.functionB` rather than just a seperate module/function? It sees it makes it unclear what is a core Faker function and what is an extension/community addition? You could also have clashes if say someone creates a "nature" module and then later we add "nature" to core.

**Comment by ST-DDT** [comment]
> But what is the advantage of having it callable via `faker.moduleA.functionB` rather than just a seperate module/function?

None, other than consistency with the other calls. It has been requested in #444.
How we will eventually achieve that, I don't want to make any predictions yet ("sub-classing", composition, ...).

> It sees it makes it unclear what is a core Faker function and what is an extension/community addition?

True, and we have to look into that. Would you consider that an important distinction and why (not)?

> You could also have clashes if say someone creates a "nature" module and then later we add "nature" to core.

The extension would win over the provided methods and thus nothing changes for the user, except that they might get additional suggestions for the native methods. In the end, it is just an outdated/obsolete dependency in your project, that you didn't remove because you didn't read the changelog/release announcement.
My hope/assumption is that all features that are eligible for inclusion will be added to Faker itself before the need arises to create an extension for that (aka the user interest required for a plugin to be published ~= the user interest required for inclusion in faker).

As a side note:

1. make it callable via faker.moduleA.functionB
2. make it able to use faker.definitions.categoryC.entryD and throw a useful error if it is missing
3. make it usable via faker.helpers.fake

1 is already possible, although only with some JS/TS tricks, that I don't want to endorse at this point.
2 is already usable.
And 3 might come for free depending on how we implement the standalone module functions (aka do we have a registry of some kind or not), well you could even achieve 1 depending on how you do/use the registry it.

Speculation: Maybe we replace faker.helpers.fake with localized function invocations (``[(fakerCore) -> `Welcome ${firstName(fakerCore)} to your exciting adventure`]``) and thus we don't need the registry to allow for better tree-shaking.
We have to keep this issue in mind when doing the v9.0 changes, but the extension system is not a v9.0 goal IMO.
